src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -130,30 +131,65 @@ private static void ModifyUnmanagedDllResolver(Action<AssemblyLoadContext> resol
-            var libExtension = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "dylib" : "so";
+            string hostFxrLibName = "libhostfxr";
+            string libExtension = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "dylib" : "so";
 
-            if (!hostFxrLibName.Equals(libraryName))
+            // the DllImport hardcoded the name as hostfxr.
+            if (!hostFxrLibName.Equals(libraryName, StringComparison.OrdinalIgnoreCase) && !libraryName.Equals("hostfxr", StringComparison.OrdinalIgnoreCase))
             {
                 return IntPtr.Zero;
             }
 
-            var hostFxrRoot = Path.Combine(DotnetPath.Value, "host", "fxr");
+            string hostFxrRoot = Path.Combine(DotnetPath.Value, "host", "fxr");
             if (Directory.Exists(hostFxrRoot))
             {
+                var fileEnumerable = new FileSystemEnumerable<SemanticVersion?>(
+                    directory: hostFxrRoot,
+                    transform: static (ref FileSystemEntry entry) => SemanticVersionParser.TryParse(entry.FileName.ToString(), out var version) ? version : null)
+                {
+                    ShouldIncludePredicate = static (ref FileSystemEntry entry) => entry.IsDirectory
+                };
+
                 // Load hostfxr from the highest version, because it should be backward-compatible
-                var hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
-                            .Max(str => SemanticVersionParser.TryParse(str, out var version) ? version : null);
+                SemanticVersion? hostFxrVersion;
+                
+                try
+                {
+                    hostFxrVersion = fileEnumerable.Max();
+                }
+                catch (ArgumentException)
+                {
+                    // LINQ Max throws when the list is empty.
+                    Console.Error.WriteLine($"No child folder was found in '{hostFxrRoot}'.");
+                    return IntPtr.Zero;
+                }
 
-                if (hostFxrAssemblyDirectory != null && !string.IsNullOrEmpty(hostFxrAssemblyDirectory.OriginalValue))
+                if (hostFxrVersion is not null)
                 {
-                    var hostFxrAssembly = Path.Combine(hostFxrAssemblyDirectory.OriginalValue, Path.ChangeExtension(hostFxrLibName, libExtension));
+                    string hostFxrAssembly = Path.Combine(hostFxrRoot, hostFxrVersion.OriginalValue, Path.ChangeExtension(hostFxrLibName, libExtension));
 
                     if (File.Exists(hostFxrAssembly))
                     {
-                        return NativeLibrary.TryLoad(hostFxrAssembly, out var handle) ? handle : IntPtr.Zero;
+                        if (NativeLibrary.TryLoad(hostFxrAssembly, out var handle))
+                        {
+                            return handle;
+                        }
+
+                        Console.Error.WriteLine($"'{hostFxrAssembly}' cannot be loaded.");
```

Comment:
I think it's much better if library code throws exceptions rather than prints to stderr. As far as I can tell exceptions thrown here should bubble up to the component that calls MSBuildLocator. Would that work for your scenario?

Reply:
hmm... somehow I got an impression that this library actually logs errors in this way, but must be a mistake. It is likely from another place related to SDK resolution.  I didn't try to throw, because I wonder this can increase risks as this method is chained to global unmanaged dll loading event.  It matches the logic to use TryLoad, instead of Load. But on the other hand, it can throw anyway from DotnetPath.Value, so I guess adopting throwing error is reasonable.

it would be nice if this code can have some Trace logging logic. But no reason to expand it further at this point.